File: frontend/src/types/notion-integration.ts
/**
 * Notion Integration Types
 * TypeScript interfaces for Notion integration functionality
 */

// Enums
export enum SyncDirection {
  NAVIMPACT_TO_NOTION = "navimpact_to_notion",
  NOTION_TO_NAVIMPACT = "notion_to_navimpact",
  BIDIRECTIONAL = "bidirectional"
}

export enum SyncStatus {
  PENDING = "pending",
  SYNCING = "syncing",
  COMPLETED = "completed",
  FAILED = "failed"
}

export enum OperationType {
  CREATE_PAGE = "create_page",
  UPDATE_PAGE = "update_page",
  DELETE_PAGE = "delete_page",
  SYNC_PROJECT = "sync_project",
  SYNC_DISTRIBUTION = "sync_distribution",
  SYNC_PERFORMANCE = "sync_performance",
  SYNC_IMPACT_STORY = "sync_impact_story"
}

export enum EntityType {
  MEDIA_PROJECT = "media_project",
  DISTRIBUTION_LOG = "distribution_log",
  PERFORMANCE_METRICS = "performance_metrics",
  IMPACT_STORY = "impact_story",
  CLIENT_ACCESS = "client_access"
}

export enum LogStatus {
  SUCCESS = "success",
  FAILED = "failed",
  PARTIAL = "partial"
}

// Base Interfaces
export interface NotionWorkspaceBase {
  workspace_id: string;
  workspace_name: string;
  bot_user_id?: string;
  is_active: boolean;
}

export interface NotionSyncMappingBase {
  notion_page_id: string;
  notion_database_id?: string;
  sync_direction: SyncDirection;
  sync_status: SyncStatus;
}

export interface NotionSyncLogBase {
  operation_type: OperationType;
  entity_type: EntityType;
  entity_id?: string;
  status: LogStatus;
  error_message?: string;
  sync_duration_ms?: number;
}

// Create Interfaces
export interface NotionWorkspaceCreate extends NotionWorkspaceBase {
  access_token: string;
}

export interface NotionSyncMappingCreate extends NotionSyncMappingBase {
  media_project_id?: number;
}

export interface NotionSyncLogCreate extends NotionSyncLogBase {
  workspace_id?: number;
}

// Update Interfaces
export interface NotionWorkspaceUpdate {
  workspace_name?: string;
  access_token?: string;
  bot_user_id?: string;
  is_active?: boolean;
}

export interface NotionSyncMappingUpdate {
  notion_database_id?: string;
  sync_direction?: SyncDirection;
  sync_status?: SyncStatus;
}

// Response Interfaces
export interface NotionWorkspaceResponse extends NotionWorkspaceBase {
  id: number;
  created_at: string;
  updated_at: string;
}

export interface NotionSyncMappingResponse extends NotionSyncMappingBase {
  id: number;
  media_project_id?: number;
  last_sync_at?: string;
  created_at: string;
}

export interface NotionSyncLogResponse extends NotionSyncLogBase {
  id: number;
  workspace_id?: number;
  created_at: string;
}

// Specialized Interfaces
export interface NotionConnectionRequest {
  code: string;
  state?: string;
}

export interface NotionSyncRequest {
  project_id: number;
  force_sync: boolean;
  sync_direction?: SyncDirection;
}

export interface NotionTemplateRequest {
  template_name: string;
  parent_page_id: string;
  project_data?: Record<string, any>;
}

export interface NotionSyncStatus {
  workspace_id: number;
  workspace_name: string;
  is_connected: boolean;
  last_sync_at?: string;
  sync_status: SyncStatus;
  total_mappings: number;
  pending_mappings: number;
  failed_mappings: number;
  recent_errors: string[];
}

export interface NotionSyncDashboard {
  workspaces: NotionWorkspaceResponse[];
  sync_status: NotionSyncStatus;
  recent_logs: NotionSyncLogResponse[];
  sync_analytics: Record<string, any>;
}

// Filter Interfaces
export interface NotionWorkspaceFilters {
  is_active?: boolean;
  workspace_name?: string;
}

export interface NotionSyncMappingFilters {
  media_project_id?: number;
  sync_status?: SyncStatus;
  sync_direction?: SyncDirection;
}

export interface NotionSyncLogFilters {
  workspace_id?: number;
  operation_type?: OperationType;
  entity_type?: EntityType;
  status?: LogStatus;
  start_date?: string;
  end_date?: string;
}

// Pagination Interfaces
export interface PaginatedNotionWorkspaces {
  items: NotionWorkspaceResponse[];
  total: number;
  page: number;
  size: number;
  pages: number;
}

export interface PaginatedNotionSyncMappings {
  items: NotionSyncMappingResponse[];
  total: number;
  page: number;
  size: number;
  pages: number;
}

export interface PaginatedNotionSyncLogs {
  items: NotionSyncLogResponse[];
  total: number;
  page: number;
  size: number;
  pages: number;
}

// Notion API Response Interfaces
export interface NotionPage {
  id: string;
  url: string;
  created_time: string;
  last_edited_time: string;
  archived: boolean;
  properties: Record<string, any>;
  parent: {
    type: string;
    page_id?: string;
    database_id?: string;
  };
}

export interface NotionDatabase {
  id: string;
  url: string;
  title: NotionRichText[];
  description: NotionRichText[];
  properties: Record<string, NotionProperty>;
  parent: {
    type: string;
    page_id?: string;
    database_id?: string;
  };
}

export interface NotionRichText {
  type: string;
  text: {
    content: string;
    link?: {
      url: string;
    };
  };
  annotations?: {
    bold: boolean;
    italic: boolean;
    strikethrough: boolean;
    underline: boolean;
    code: boolean;
    color: string;
  };
  plain_text: string;
  href?: string;
}

export interface NotionProperty {
  id: string;
  type: string;
  name: string;
  title?: {
    rich_text: NotionRichText[];
  };
  rich_text?: {
    rich_text: NotionRichText[];
  };
  number?: number;
  select?: {
    name: string;
    color: string;
  };
  multi_select?: Array<{
    name: string;
    color: string;
  }>;
  date?: {
    start: string;
    end?: string;
  };
  checkbox?: boolean;
  url?: string;
  email?: string;
  phone_number?: string;
  formula?: {
    type: string;
    string?: string;
    number?: number;
    boolean?: boolean;
    date?: {
      start: string;
      end?: string;
    };
  };
  relation?: Array<{
    id: string;
  }>;
  rollup?: {
    type: string;
    number?: number;
    date?: {
      start: string;
      end?: string;
    };
    array?: any[];
  };
  people?: Array<{
    object: string;
    id: string;
    name?: string;
    avatar_url?: string;
    type: string;
    person?: {
      email: string;
    };
    bot?: {
      workspace_name: string;
    };
  }>;
  files?: Array<{
    name: string;
    type: string;
    file?: {
      url: string;
      expiry_time: string;
    };
    external?: {
      url: string;
    };
  }>;
  created_by?: {
    object: string;
    id: string;
  };
  created_time?: string;
  last_edited_by?: {
    object: string;
    id: string;
  };
  last_edited_time?: string;
}

// Template Interfaces
export interface NotionTemplate {
  id: string;
  name: string;
  description: string;
  category: string;
  properties: Record<string, NotionProperty>;
  content: NotionBlock[];
}

export interface NotionBlock {
  object: string;
  id: string;
  type: string;
  created_time: string;
  last_edited_time: string;
  has_children: boolean;
  archived: boolean;
  heading_1?: {
    rich_text: NotionRichText[];
    color: string;
    is_toggleable: boolean;
  };
  heading_2?: {
    rich_text: NotionRichText[];
    color: string;
    is_toggleable: boolean;
  };
  heading_3?: {
    rich_text: NotionRichText[];
    color: string;
    is_toggleable: boolean;
  };
  paragraph?: {
    rich_text: NotionRichText[];
    color: string;
  };
  bulleted_list_item?: {
    rich_text: NotionRichText[];
    color: string;
  };
  numbered_list_item?: {
    rich_text: NotionRichText[];
    color: string;
  };
  callout?: {
    rich_text: NotionRichText[];
    icon: {
      type: string;
      emoji?: string;
      file?: {
        url: string;
        expiry_time: string;
      };
      external?: {
        url: string;
      };
    };
    color: string;
  };
  table_of_contents?: {
    color: string;
  };
  divider?: {
    color: string;
  };
  image?: {
    caption: NotionRichText[];
    type: string;
    file?: {
      url: string;
      expiry_time: string;
    };
    external?: {
      url: string;
    };
  };
}

// Component Props Interfaces
export interface NotionConnectionProps {
  onConnect: (workspace: NotionWorkspaceResponse) => void;
  onError: (error: string) => void;
}

export interface NotionSyncStatusProps {
  syncStatus: NotionSyncStatus;
  onRefresh: () => void;
}

export interface NotionSyncLogProps {
  logs: NotionSyncLogResponse[];
  onViewDetails: (log: NotionSyncLogResponse) => void;
}

export interface NotionTemplateSelectorProps {
  templates: NotionTemplate[];
  onSelectTemplate: (template: NotionTemplate) => void;
  selectedTemplate?: NotionTemplate;
}

export interface NotionSyncControlsProps {
  projectId: number;
  onSync: (request: NotionSyncRequest) => void;
  isSyncing: boolean;
  lastSyncAt?: string;
} 

File: app/schemas/notion_integration.py
"""
Notion Integration Schemas
Pydantic models for Notion integration API requests and responses
"""

from datetime import datetime
from enum import Enum
from typing import Any, Dict, List, Optional

from pydantic import BaseModel, Field


class SyncDirection(str, Enum):
    """Sync direction options."""

    NAVIMPACT_TO_NOTION = "navimpact_to_notion"
    NOTION_TO_NAVIMPACT = "notion_to_navimpact"
    BIDIRECTIONAL = "bidirectional"


class SyncStatus(str, Enum):
    """Sync status options."""

    PENDING = "pending"
    SYNCING = "syncing"
    COMPLETED = "completed"
    FAILED = "failed"


class OperationType(str, Enum):
    """Operation type options."""

    CREATE_PAGE = "create_page"
    UPDATE_PAGE = "update_page"
    DELETE_PAGE = "delete_page"
    SYNC_PROJECT = "sync_project"
    SYNC_DISTRIBUTION = "sync_distribution"
    SYNC_PERFORMANCE = "sync_performance"
    SYNC_IMPACT_STORY = "sync_impact_story"


class EntityType(str, Enum):
    """Entity type options."""

    MEDIA_PROJECT = "media_project"
    DISTRIBUTION_LOG = "distribution_log"
    PERFORMANCE_METRICS = "performance_metrics"
    IMPACT_STORY = "impact_story"
    CLIENT_ACCESS = "client_access"


class LogStatus(str, Enum):
    """Log status options."""

    SUCCESS = "success"
    FAILED = "failed"
    PARTIAL = "partial"


# Base Models
class NotionWorkspaceBase(BaseModel):
    """Base Notion workspace model."""

    workspace_id: str = Field(..., description="Notion workspace ID")
    workspace_name: str = Field(..., description="Notion workspace name")
    bot_user_id: Optional[str] = Field(None, description="Notion bot user ID")
    is_active: bool = Field(True, description="Whether the workspace is active")


class NotionSyncMappingBase(BaseModel):
    """Base Notion sync mapping model."""

    notion_page_id: str = Field(..., description="Notion page ID")
    notion_database_id: Optional[str] = Field(None, description="Notion database ID")
    sync_direction: SyncDirection = Field(
        SyncDirection.BIDIRECTIONAL, description="Sync direction"
    )
    sync_status: SyncStatus = Field(
        SyncStatus.PENDING, description="Current sync status"
    )


class NotionSyncLogBase(BaseModel):
    """Base Notion sync log model."""

    operation_type: OperationType = Field(..., description="Type of operation")
    entity_type: EntityType = Field(..., description="Type of entity")
    entity_id: Optional[str] = Field(None, description="Entity ID")
    status: LogStatus = Field(..., description="Operation status")
    error_message: Optional[str] = Field(None, description="Error message if failed")
    sync_duration_ms: Optional[int] = Field(
        None, description="Sync duration in milliseconds"
    )


# Create Models
class NotionWorkspaceCreate(NotionWorkspaceBase):
    """Create Notion workspace model."""

    access_token: str = Field(..., description="Notion access token")


class NotionSyncMappingCreate(NotionSyncMappingBase):
    """Create Notion sync mapping model."""

    media_project_id: Optional[int] = Field(None, description="SGE Media Project ID")


class NotionSyncLogCreate(NotionSyncLogBase):
    """Create Notion sync log model."""

    workspace_id: Optional[int] = Field(None, description="Notion workspace ID")


# Update Models
class NotionWorkspaceUpdate(BaseModel):
    """Update Notion workspace model."""

    workspace_name: Optional[str] = Field(None, description="Notion workspace name")
    access_token: Optional[str] = Field(None, description="Notion access token")
    bot_user_id: Optional[str] = Field(None, description="Notion bot user ID")
    is_active: Optional[bool] = Field(
        None, description="Whether the workspace is active"
    )


class NotionSyncMappingUpdate(BaseModel):
    """Update Notion sync mapping model."""

    notion_database_id: Optional[str] = Field(None, description="Notion database ID")
    sync_direction: Optional[SyncDirection] = Field(None, description="Sync direction")
    sync_status: Optional[SyncStatus] = Field(None, description="Current sync status")


# Response Models
class NotionWorkspaceResponse(NotionWorkspaceBase):
    """Notion workspace response model."""

    id: int
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True


class NotionSyncMappingResponse(NotionSyncMappingBase):
    """Notion sync mapping response model."""

    id: int
    media_project_id: Optional[int]
    last_sync_at: Optional[datetime]
    created_at: datetime

    class Config:
        from_attributes = True


class NotionSyncLogResponse(NotionSyncLogBase):
    """Notion sync log response model."""

    id: int
    workspace_id: Optional[int]
    created_at: datetime

    class Config:
        from_attributes = True


# Specialized Models
class NotionConnectionRequest(BaseModel):
    """Notion connection request model."""

    code: str = Field(..., description="Authorization code from Notion OAuth")
    state: Optional[str] = Field(None, description="State parameter for security")


class NotionSyncRequest(BaseModel):
    """Notion sync request model."""

    project_id: int = Field(..., description="SGE Media Project ID to sync")
    force_sync: bool = Field(False, description="Force sync even if recently synced")
    sync_direction: Optional[SyncDirection] = Field(
        None, description="Override sync direction"
    )


class NotionTemplateRequest(BaseModel):
    """Notion template request model."""

    template_name: str = Field(..., description="Name of the template to create")
    parent_page_id: str = Field(
        ..., description="Parent page ID where template will be created"
    )
    project_data: Optional[Dict[str, Any]] = Field(
        None, description="Project data to populate template"
    )


class NotionSyncStatus(BaseModel):
    """Notion sync status response model."""

    workspace_id: int
    workspace_name: str
    is_connected: bool
    last_sync_at: Optional[datetime]
    sync_status: SyncStatus
    total_mappings: int
    pending_mappings: int
    failed_mappings: int
    recent_errors: List[str] = Field(default_factory=list)


class NotionSyncDashboard(BaseModel):
    """Notion sync dashboard response model."""

    workspaces: List[NotionWorkspaceResponse]
    sync_status: NotionSyncStatus
    recent_logs: List[NotionSyncLogResponse]
    sync_analytics: Dict[str, Any] = Field(default_factory=dict)


# Filter Models
class NotionWorkspaceFilters(BaseModel):
    """Notion workspace filters."""

    is_active: Optional[bool] = Field(None, description="Filter by active status")
    workspace_name: Optional[str] = Field(None, description="Filter by workspace name")


class NotionSyncMappingFilters(BaseModel):
    """Notion sync mapping filters."""

    media_project_id: Optional[int] = Field(
        None, description="Filter by media project ID"
    )
    sync_status: Optional[SyncStatus] = Field(None, description="Filter by sync status")
    sync_direction: Optional[SyncDirection] = Field(
        None, description="Filter by sync direction"
    )


class NotionSyncLogFilters(BaseModel):
    """Notion sync log filters."""

    workspace_id: Optional[int] = Field(None, description="Filter by workspace ID")
    operation_type: Optional[OperationType] = Field(
        None, description="Filter by operation type"
    )
    entity_type: Optional[EntityType] = Field(None, description="Filter by entity type")
    status: Optional[LogStatus] = Field(None, description="Filter by status")
    start_date: Optional[datetime] = Field(None, description="Filter by start date")
    end_date: Optional[datetime] = Field(None, description="Filter by end date")


# Pagination Models
class PaginatedNotionWorkspaces(BaseModel):
    """Paginated Notion workspaces response."""

    items: List[NotionWorkspaceResponse]
    total: int
    page: int
    size: int
    pages: int


class PaginatedNotionSyncMappings(BaseModel):
    """Paginated Notion sync mappings response."""

    items: List[NotionSyncMappingResponse]
    total: int
    page: int
    size: int
    pages: int


class PaginatedNotionSyncLogs(BaseModel):
    """Paginated Notion sync logs response."""

    items: List[NotionSyncLogResponse]
    total: int
    page: int
    size: int
    pages: int
